src/NHibernate/Hql/Ast/ANTLR/Tree/JoinSubqueryFromElement.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System.Linq;
+using Antlr.Runtime;
+using NHibernate.Engine;
+using NHibernate.Persister.Entity;
+using NHibernate.SqlCommand;
+using NHibernate.Type;
+
+namespace NHibernate.Hql.Ast.ANTLR.Tree
+{
+	internal class JoinSubqueryFromElement : FromElement
+	{
+		public JoinSubqueryFromElement(FromClause fromClause, QueryNode queryNode, JoinType joinType, string alias)
+			: base(new CommonToken(HqlSqlWalker.JOIN_SUBQUERY, "join sub-query"))
+		{
+			var tableAlias = fromClause.AliasGenerator.CreateName("subquery");
+			var querySelectClause = queryNode.GetSelectClause();
+			var dataType = CreateDataType(querySelectClause, out var entityPersister);
+			PropertyMapping = new SubqueryPropertyMapping(dataType, querySelectClause);
+			QueryNode = queryNode;
+			DataType = dataType;
+			foreach (var fromElement in querySelectClause.SelectExpressions.Select(o => o.FromElement)
+				.Union(querySelectClause.NonScalarExpressions?.Select(o => o.FromElement) ?? Enumerable.Empty<FromElement>())
+				.Union(querySelectClause.CollectionFromElements ?? Enumerable.Empty<FromElement>())
```

Comment:
If I'm not mistaken each `Union` call leads to `Set` creation with all existing items So it's better  to do `Concat` with single `Distinct` call in the end.

Reply:
Done as suggested.